parser-typechecker/src/Unison/Util/Bytes.hs

```diff
@@ -1,42 +1,78 @@
 {-# Language ViewPatterns #-}
+{-# Language GeneralizedNewtypeDeriving #-}
+{-# Language BangPatterns #-}
 {-# OPTIONS_GHC -fno-warn-orphans #-}
 
 module Unison.Util.Bytes where
 
 import Control.DeepSeq (NFData(..))
 import Data.Bits (shiftR, shiftL, (.|.))
 import Data.Char
-import Data.Memory.PtrMethods (memCompare, memEqual)
-import Data.Monoid (Sum(..))
-import Foreign.Ptr (Ptr, plusPtr)
-import Foreign.Storable (poke)
-import System.IO.Unsafe (unsafeDupablePerformIO)
 import Unison.Prelude hiding (ByteString, empty)
-import Basement.Block (Block)
-import qualified Data.ByteString.Lazy as LB
-import qualified Data.ByteArray as B
+import qualified Data.ByteString as B
+import qualified Data.ByteArray as BA
 import qualified Data.ByteArray.Encoding as BE
-import qualified Data.FingerTree as T
+import qualified Data.ByteString.Lazy as LB
 import qualified Data.Text as Text
+import qualified Unison.Util.Rope as R
+import qualified Data.Vector.Primitive as V
+import qualified Data.Vector.Storable as SV
+import qualified Data.Vector.Storable.ByteString as BSV
+import Foreign.Storable (pokeByteOff)
 import qualified Codec.Compression.Zlib as Zlib
 import qualified Codec.Compression.GZip as GZip
 
--- Block is just `newtype Block a = Block ByteArray#`
-type ByteString = Block Word8
+type ByteString = V.Vector Word8
+
+-- Bytes type represented as a rope of ByteStrings
+newtype Bytes = Bytes { underlying :: R.Rope ByteString } deriving (Semigroup,Monoid,Eq,Ord)
 
--- Bytes type represented as a finger tree of ByteStrings.
--- Can be efficiently sliced and indexed, using the byte count
--- annotation at each subtree.
-newtype Bytes = Bytes (T.FingerTree (Sum Int) (View ByteString))
+instance R.Sized ByteString where size = V.length
+instance R.Drop ByteString where drop = V.drop
+instance R.Take ByteString where take = V.take
+instance R.Index ByteString Word8 where index n bs = bs V.!? n
+instance R.Reverse ByteString where reverse = V.reverse
+instance NFData Bytes where rnf _ = ()
 
 null :: Bytes -> Bool
-null (Bytes bs) = T.null bs
+null = R.null . underlying
 
 empty :: Bytes
-empty = Bytes mempty
+empty = mempty
+
+isAscii :: Bytes -> Bool
```

Comment:
Doesn't have to be as part of this PR of course, but it would likely be good to add time complexities to the docs for each of these so it's easy to see at a glance what performance we can expect when using them, e.g. looks like this is `O(n)`